Verifier split

exFalso opened the pull request:
This PR adds a couple of things:

* Hides verification behind a new `TransactionVerifierService`
* Adds an InMemory verifier (used by default) backed by a worker pool
* Adds an out of process verifier (in a new module)
* Adds yet another module `node-api`. The intent is to factor out shared code between the node and components interfacing with the node. I only added minimal code here required for the verifier, next would be RPC, which would allow the breaking of `:node` dependencies throughout the code.
* Adds configurable certificate checking policies per artemis role. The reasoning behind this is we don't know how end users want to do cert checks for different node components (which we may/will have several of in the future)
* Adds `VerifierDriver` extending `Driver` allowing starting of verifiers and lightweight verification requestors.
* Adds a valid ledger generator

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -183,10 +184,11 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
     /**
-     * Authenticated clients connecting to us fall in one of three groups:
+     * Authenticated clients connecting to us fall in one of the following groups:
      * 1. The node hosting us and any of its logically connected components. These are given full access to all valid queues.
```
This should just mention node and not any of its components, as the verifier could be counted as one.

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -453,65 +536,74 @@ class NodeLoginModule : LoginModule {
+
+            loginSucceeded = true
+            return loginSucceeded
+        } catch (exception: FailedLoginException) {
```
Not sure we want to do this as the docs for `LoginModule.login` says it can throw `LoginException` if the authentication fails.

exFalso replied:
I added this because these exceptions are completely swallowed otherwise. Should I rethrow?

shamsasari replied:
I imagine Artemis must be logging the exceptions thrown somewhere - perhaps we just need to find where and manually enable that log.

shamsasari replied:
If you want to be doing the logging here and not rely on Artemis, then rethrow.

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/NodeMessagingClient.kt:
```diff
@@ -154,6 +171,18 @@ class NodeMessagingClient(override val config: NodeConfiguration,
+                    log.warn("No connected verifier listening on $VERIFICATION_REQUESTS_QUEUE_NAME!")
+                }
+            }
+            if (config.verifierType == VerifierType.OutOfProcess) {
```
Add blank line before the if

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/NodeMessagingClient.kt:
```diff
@@ -154,6 +171,18 @@ class NodeMessagingClient(override val config: NodeConfiguration,
+                session.createQueue(VerifierApi.VERIFICATION_REQUESTS_QUEUE_NAME, VerifierApi.VERIFICATION_REQUESTS_QUEUE_NAME)
+                session.createQueue(VERIFIER_RESPONSE_ADDRESS, VERIFIER_RESPONSE_ADDRESS)
+                verificationResponseConsumer = session.createConsumer(VERIFIER_RESPONSE_ADDRESS)
+                messagingExecutor.scheduleAtFixedRate(::checkVerifierCount, 0, 10, TimeUnit.SECONDS)
```
Could we use Artemis' management feature to do the same thing? https://activemq.apache.org/artemis/docs/1.5.3/management.html

> CONSUMER_CLOSED (3)

> _AMQ_Address, _AMQ_ClusterName, _AMQ_RoutingName, _AMQ_Distance, _AMQ_ConsumerCount,
> _AMQ_User, _AMQ_RemoteAddress, _AMQ_SessionName, _AMQ_FilterString

exFalso replied:
I believe `queueQuery` does just that

shamsasari commented on node/src/main/kotlin/net/corda/node/services/transactions/OutOfProcessTransactionVerifierService.kt:
```diff
@@ -0,0 +1,40 @@
+    override fun verify(transaction: LedgerTransaction): ListenableFuture<*> {
+        val future = SettableFuture.create<Unit>()
+        val nonce = random63BitValue()
+        verificationResponseFutures.set(nonce, future)
```
`verificationResponseFutures[nonce] = future`

shamsasari commented on node/src/test/kotlin/net/corda/node/services/MockServiceHubInternal.kt:
```diff
@@ -33,8 +34,11 @@ open class MockServiceHubInternal(
         val flowFactory: FlowLogicRefFactory? = FlowLogicRefFactory(),
-        val schemas: SchemaService? = NodeSchemaService()
+        val schemas: SchemaService? = NodeSchemaService(),
+        val customTransactionVerifierService: TransactionVerifierService? = InMemoryTransactionVerifierService(2)
```
What does the custom one do?

exFalso replied:
It can be used to pass in e.g. a mock verifier service. I tried to be consistent with the existing code here

shamsasari replied:
Hmm, I think this needs a cleanup at some point.

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/GeneratedLedger.kt:
```diff
@@ -0,0 +1,268 @@
+        val empty = GeneratedLedger(emptyList(), emptyMap(), emptySet(), emptySet())
+    }
+
+    fun resolveWireTransaction(transaction: WireTransaction): LedgerTransaction {
```
This method's a bit big - extract out the anonymous classes and give them names.

exFalso replied:
Good point. I initially changed `toLedgerTransaction` to just accept lookup functions instead of a full service interface, do you think that'd be better?

shamsasari commented on verifier/src/main/kotlin/net/corda/verifier/Verifier.kt:
```diff
@@ -0,0 +1,83 @@
+    companion object {
+        private val log = loggerFor<Verifier>()
+
+        fun loadConfiguration(baseDirectory: String): VerifierConfiguration {
```
Use `Path`

shamsasari commented on verifier/src/main/kotlin/net/corda/verifier/Verifier.kt:
```diff
@@ -0,0 +1,83 @@
+
+        @JvmStatic
+        fun main(args: Array<String>) {
+            if (args.isEmpty()) {
```
Use `require`

shamsasari commented on verifier/src/main/kotlin/net/corda/verifier/Verifier.kt:
```diff
@@ -0,0 +1,83 @@
+            val replyProducer = session.createProducer()
+            consumer.setMessageHandler {
+                val request = VerifierApi.VerificationRequest.fromClientMessage(it)
+                log.debug("Received verification request with id ${request.verificationId}")
```
Use the extension method: `log.debug { ... }`

shamsasari commented on verifier/src/main/kotlin/net/corda/verifier/Verifier.kt:
```diff
@@ -0,0 +1,83 @@
+                    request.transaction.verify()
+                }
+                if (result.error != null) {
+                    log.debug("Verification returned with error ${result.error}")
```
Ditto

exFalso commented on node-api/src/main/kotlin/net/corda/node/VerifierApi.kt:
```diff
@@ -0,0 +1,63 @@
+    data class VerificationRequest(
+            val verificationId: Long,
+            val transaction: LedgerTransaction,
+            val responseAddress: SimpleString
```
Yes, this module is intended to capture the actual node's interface

exFalso commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -263,20 +275,73 @@ private fun <A> poll(
+    private val state = ThreadBox(State())
+
+    fun shutdown() {
+        require(!state.locked { isShutdown })
```
Very true!

exFalso commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -263,20 +275,73 @@ private fun <A> poll(
+
+    fun registerProcessShutdown(processFuture: ListenableFuture<Process>) {
+        val processShutdown = processFuture.map { process ->
+            {
```
We do, the future returns a closure

shamsasari replied:
It would clearer if this was extracted out into a method

exFalso commented on core/src/main/kotlin/net/corda/core/node/services/Services.kt:
```diff
@@ -322,3 +323,10 @@ interface SchedulerService {
+ * Provides verification service. The implementation may be a simple in-memory verify() call or perhaps an IPC/RPC.
+ */
+interface TransactionVerifierService {
+    fun verify(transaction: LedgerTransaction): ListenableFuture<*>
```
This got me thinking, currently if an out of process verifier dies during verification the future will just hang, no exceptions will be set here. I'll think about this a bit more

mikehearn commented:
@shamsasari let me know when you've kicked the review ball into my court here.

shamsasari commented:
Sure. I'm just waiting for @exFalso's OK for another round.

exFalso commented:
I'm adding some additional tests that kill verifiers, it works, but for some reason it results in a large number of dupes, need to debug that

exFalso commented:
Rebased, comments addressed. The duping problem was due to incorrect acknowledgement commits

shamsasari commented on config/dev/log4j2.xml:
```diff
@@ -5,8 +5,10 @@
-        <Property name="defaultLogLevel">info</Property>
+        <Property name="consoleLogLevel">info
+        </Property>
+        <Property name="defaultLogLevel">trace
```
Did you forget to revert this?

shamsasari commented on core/src/main/kotlin/net/corda/core/node/services/Services.kt:
```diff
@@ -322,3 +323,15 @@ interface SchedulerService {
+    /**
+     * @param transaction The transaction to be verified.
+     * @return A future that completes successfully if the transaction verified, or sets an exception the verifier threw.
+     * @throws Throwable
```
Give more details or remove.

shamsasari commented on core/src/main/kotlin/net/corda/core/utilities/Logging.kt:
```diff
@@ -21,6 +21,10 @@ inline fun org.slf4j.Logger.debug(msg: () -> String) {
     if (isDebugEnabled) debug(msg())
 }
 
+inline fun org.slf4j.Logger.info(msg: () -> String) {
```
Info is usually on in production, so I doubt we will benefit from this optimisation.

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/NodeMessagingClient.kt:
```diff
@@ -461,6 +492,23 @@ class NodeMessagingClient(override val config: NodeConfiguration,
+        return object : OutOfProcessTransactionVerifierService() {
+                override fun sendRequest(nonce: Long, transaction: LedgerTransaction) {
+                    messagingExecutor.fetchFrom {
+                        state.locked {
```
I wonder if we're risking blocking the executor here while we wait for the lock. Should it be done inside just around the send?

exFalso replied:
I believe any session access must be guarded, artemis is quite allergic to concurrent access

shamsasari replied:
Ah yes of course.

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/VerifierDriver.kt:
```diff
@@ -0,0 +1,288 @@
+            val consumer = session.createConsumer(responseAddress)
+            // We demux the individual txs ourselves to avoid race when a new verifier is added
+            val verificationResponseFutures = ConcurrentHashMap<Long, SettableFuture<Throwable?>>()
+            val asd = AtomicInteger(0)
```
asd?

exFalso replied:
Some remains of debugging, thank you!

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/VerifierTests.kt:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Test
+    fun singleVerifierWorksWithRequestor() {
```
Consider using back ticks so that we can have spaces between the words - makes the test names clearer

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/VerifierDriver.kt:
```diff
@@ -0,0 +1,288 @@
+
+    override fun startVerificationRequestor(name: String): ListenableFuture<VerificationRequestorHandle> {
+        val hostAndPort = driverDSL.portAllocation.nextHostAndPort()
+        return driverDSL.executorService.submit<VerificationRequestorHandle> {
```
This lambda's far too big. You could probably extract out the most of this code in a `VerificationRequestor` class and keep the register calls here.

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/VerifierTests.kt:
```diff
@@ -0,0 +1,125 @@
+    @Test
+    fun singleVerifierWorksWithRequestor() {
+        verifierDriver(automaticallyStartNetworkMap = false) {
+            val aliceFuture = startVerificationRequestor("Alice")
```
Do we need to give the requesters names? It gives the impression they're nodes.

exFalso replied:
We just need a unique directory for each, this way it's easy to identify which one's which.

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/VerifierTests.kt:
```diff
@@ -0,0 +1,125 @@
+            val numberOfTransactions = 100
+            val transactions = generateTransactions(numberOfTransactions)
+            val alice = aliceFuture.get()
+            val a = startVerifier(alice)
```
verifier1, verifier2, verifier3

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/VerifierTests.kt:
```diff
@@ -0,0 +1,125 @@
+            val b = startVerifier(alice)
+            val c = startVerifier(alice)
+            alice.waitUntilNumberOfVerifiers(3)
+            val remainingCounter = AtomicInteger(numberOfTransactions)
```
remainingTxCount

exFalso commented:
Back to you

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/VerifierDriver.kt:
```diff
@@ -0,0 +1,292 @@
+        }
+    }
+
+    private fun startVerificationRequestorInternal(name: String, hostAndPort: HostAndPort): VerificationRequestorHandle {
```
The big lambda has become a big method!

shamsasari commented on core/src/main/kotlin/net/corda/flows/ResolveTransactionsFlow.kt:
```diff
@@ -107,7 +108,9 @@ class ResolveTransactionsFlow(private val txHashes: Set<SecureHash>,
-            ltx.verify()
+            // Block on each verification request.
+            // TODO We could recover some parallelism from the dependency graph.
+            serviceHub.transactionVerifierService.verify(ltx).getOrThrow()
```
I just realised we should actually be suspending the fiber here until the response comes back, rather than blocking. This seems to go into the realm of having flows suspend until an external notification happens.

mikehearn replied:
Yes, but previously verification was blocking as well. So it's no downgrade. Figuring out how to integrate futures with the flow framework is a task for a future PR (quasar already does it easily but the question is checkpointing semantics).

shamsasari commented:
Also what's to prevent a flow developer from bypassing the verifier and simply call `LedgerTransaction.verify()`?

exFalso commented:
Do you think we should safeguard against it?

shamsasari commented:
My gut feeling is that even if we provide good clear documentation on how to verify transactions, someone will see that method (it is public after all) and realise it's much more "convenient" to use that instead of fiddling around with futures and stuff.

mikehearn commented:
Why can't we just make LedgerTransaction.verify invoke the right flow when it's in flow context automatically? Or, failing that, throw with an exception that tells the developer what to do.

But I bet if we can do the latter we can do the former. Rough guess at an implementation - change LedgerTransaction.verify to take a verifier interface that defaults to the contents of a  TLS slot or a direct implementation when the slot is empty. Flow framework points the slot back at itself to trigger the subflow. LedgerTransaction.verify and intermediates are marked as suspendable.

mnesbit commented on node/src/main/kotlin/net/corda/node/internal/Node.kt:
```diff
@@ -56,6 +56,7 @@ class Node(override val configuration: FullNodeConfiguration,
     override val version: Version get() = nodeVersionInfo.version
     override val networkMapAddress: NetworkMapAddress? get() = configuration.networkMapService?.address?.let(::NetworkMapAddress)
+    override fun makeTransactionVerifierService() = (net as NodeMessagingClient).verifierService
 
```
What is the correct init point for this? especially if I start my node before my the verifier? Is it safe to start my statemachine, or will it break?

exFalso replied:
Verifiers may be started before or after. There is a test for starting the verifier after the node.

I also added a polling method that prints warnings if there is are no connected verifiers. In these cases verification will hang until a verifier comes online.

exFalso replied:
Hold on that's incorrect, verifiers must start after the node has started, they are artemis clients

chrisr3 commented on verifier/build.gradle:
```diff
@@ -0,0 +1,115 @@
+    testCompile project(":node")
+    testCompile project(":test-utils")
+
+    compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
```
Aren't all of these Kotlin dependencies inherited transitively from `core`?

chrisr3 commented on node-api/build.gradle:
```diff
@@ -0,0 +1,78 @@
+    testCompile project(":node")
+    testCompile project(":test-utils")
+
+    compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
```
These Kotlin dependencies should all be transitive dependencies of `core`. I'm not sure why `core` wants to include Kotlin's test framework at compile scope, though...

mikehearn commented:
We always specify dependencies directly, not transitively inherit them. For one, dependencies of dependencies are not stable API so you can't rely on them sticking around. For another it's required as we evolve towards the java module system.

mikehearn commented on core/src/main/kotlin/net/corda/core/Utils.kt:
```diff
@@ -102,7 +102,8 @@ fun <T> ListenableFuture<T>.failure(executor: Executor, body: (Throwable) -> Uni
 infix fun <T> ListenableFuture<T>.failure(body: (Throwable) -> Unit): ListenableFuture<T> = apply { failure(RunOnCallerThread, body) }
-infix fun <F, T> ListenableFuture<F>.map(mapper: (F) -> T): ListenableFuture<T> = Futures.transform(this, Function { mapper(it!!) })
+@Suppress("UNCHECKED_CAST") // We need the awkward cast because otherwise F cannot be nullable, even though it's safe.
+infix fun <F, T> ListenableFuture<F>.map(mapper: (F) -> T): ListenableFuture<T> = Futures.transform(this, Function { (mapper as (F?) -> T)(it) })
```
Can't you just make the type variable nullable?

`infix fun <F, T> ListenableFuture<F?>.map(mapper: (F?) -> T): ListenableFuture<T?> = ....`

exFalso replied:
In that case we would only be able to use it on nullable futures

mikehearn replied:
This code compiles:

```
infix fun <F : Any?, T> ListenableFuture<F>.map(mapper: (F?) -> T): ListenableFuture<T> = Futures.transform(this, Function { mapper(it!!) })

fun foo(x: ListenableFuture<String>) {
    x.map { it!!.reversed() }
}
```

Isn't that sort of what we want? We just need to relax the upper bound on the type variable.

shamsasari replied:
`String` is not nullable but we still need to do `!!`? That's not a friendly API.

exFalso replied:
^ what Shams said. The cast looks weird, but it's actually safe, the only reason `?` creeps in is because in guava the param is annotated with @Nullable

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt:
```diff
@@ -71,16 +71,29 @@ class WireTransaction(
+    }
+
+    @Throws(AttachmentResolutionException::class, TransactionResolutionException::class)
+    fun toLedgerTransaction(
```
This should either be private, or have a kdoc.

chrisr3 commented:
> We always specify dependencies directly, not transitively inherit them.
Almost all of the transitive dependencies that I am referring to are already explicitly specified in our own `core` module, and are therefore stable.

exFalso commented:
Split this into #379, #381, rebased this one onto #379 and merged #381. Will remove the merge commit once 381 is merged into master

mikehearn commented:
@exFalso let me know when it's time to re-review.

exFalso commented:
This PR is now unblocked.

Rebased and addressed latest comments: added kdocs to `toLedgerTransaction` overload, removed unnecessary deps in build.gradle.

I did not address the user-friendlier `verify()` suggestion, I think this requires some more thought and doesn't belong in this PR. I would like to avoid adding flow-specific functionality to LedgerTransaction as we'll need to use it in non-flow contexts as well where we can't even refer to flow-specific types (for example in the verifier itself)

shamsasari commented on docs/source/out-of-process-verification.rst:
```diff
@@ -0,0 +1,20 @@
+
+A Corda node does transaction verification through ``ServiceHub.transactionVerifierService``. This is by default an ``InMemoryTransactionVerifierService`` which just verifies transactions in-process.
+
+Corda may be configured to use out of process verification. Any number of verifiers may be started connecting to a node through the node's exposed artemis SSL port. The messaging layer takes care of loadbalancing.
```
Typo: load balancing

shamsasari commented on node/src/integration-test/kotlin/net/corda/services/messaging/MQSecurityTest.kt:
```diff
@@ -2,7 +2,11 @@ package net.corda.services.messaging
 
 import co.paralleluniverse.fibers.Suspendable
 import com.google.common.net.HostAndPort
+<<<<<<< HEAD
```
Merge issues?

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/VerifierDriver.kt:
```diff
@@ -0,0 +1,297 @@
+        writeConfig(baseDirectory, configFilename, config)
+        Verifier.loadConfiguration(baseDirectory, baseDirectory / configFilename).configureDevKeyAndTrustStores(verifierName)
+
+        val className = Verifier::class.java.name
```
A lot of this start-a-java-process code is duplicated; now 3 times.

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/GeneratedLedger.kt:
```diff
@@ -0,0 +1,230 @@
+
+    /**
+     * Generates an issuance(root) transaction.
+     * Invariants: The input list must be empty.
```
You mean the input lists of `transactions`? If so then add a `require` check.

exFalso replied:
No, this is not a precondition, it just states the invariants of the generated transaction

shamsasari replied:
I see. I guess the use of the word "must" implied it was a pre-condition.

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/GeneratedLedger.kt:
```diff
@@ -0,0 +1,230 @@
+    /**
+     * Generates a regular non-issue transaction.
+     * Invariants:
+     *   * Input and output notaries must be one and the same.
```
Likewise add checks for this

exFalso replied:
See previous comment

shamsasari commented on verifier/src/integration-test/kotlin/net/corda/verifier/GeneratedLedger.kt:
```diff
@@ -0,0 +1,230 @@
+ * It provides generator methods, in particular [transactionGenerator] that generates a valid transaction and also
+ * returns the new state of the ledger.
+ */
+data class GeneratedLedger(
```
This doesn't seem to be used anywhere. And there are no usage examples. Perhaps move this onto another PR if you have plans for this.

exFalso replied:
It's used in the verifier integration tests

exFalso commented:
Back to you

mikehearn commented on node/src/main/kotlin/net/corda/node/services/transactions/OutOfProcessTransactionVerifierService.kt:
```diff
@@ -0,0 +1,42 @@
+        val future = SettableFuture.create<Unit>()
+        val nonce = random63BitValue()
+        verificationResponseFutures[nonce] = future
+        sendRequest(nonce, transaction)
```
Can we add some logging and some timing metrics here?

exFalso commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -447,6 +459,66 @@ private class VerifyingNettyConnector(configuration: MutableMap<String, Any>?,
+            val rootPublicKey = trustStore.getCertificate(CORDA_ROOT_CA).publicKey
+            return object : Check {
+                override fun checkCertificateChain(theirChain: Array<X509Certificate>) {
+                    val theirRoot = theirChain.last().publicKey
```
That was the previous behaviour, should I change it?

mikehearn replied:
I see, OK, then leave it for now.

exFalso commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -447,6 +459,66 @@ private class VerifyingNettyConnector(configuration: MutableMap<String, Any>?,
     }
 }
 
+sealed class CertificateChainCheckPolicy {
```
The reason I introduced this was so it can be configured from node.conf, so it needs to be visible from there. internal?

exFalso commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -447,6 +459,66 @@ private class VerifyingNettyConnector(configuration: MutableMap<String, Any>?,
+sealed class CertificateChainCheckPolicy {
+
+    @FunctionalInterface
+    interface Check {
```
That's what I went with first, but the types became a little verbose:
```
abstract fun createCheck(keyStore: KeyStore, trustStore: KeyStore): (Array<X509Certificate>)->Unit
```
so I figured this is more readable

mikehearn replied:
Sounds like something typealias in Kotlin 1.1 would address but it's fine as is.

exFalso commented:
Rebased, should be ready for re-review
